### ISSUE TITLE
Certain base functions on JuMP constraints not implemented for NLconstraints.

### ISSUE THREAD

**Issue by 1ozturkbe** [jump-dev/JuMP.jl#2355]
Hey folks, I'm struggling a bit since there are certain basic methods on JuMP constraints that are not defined with NLconstraints, eg.:

``` julia
delete(::JuMP.Model, ::NonlinearConstraintIndex)
is_valid(::JuMP.Model, ::ConstraintRef{Model, NonlinearConstraintIndex, ScalarShape})
```
simply do not work. I was wondering what the design behind this is, and whether there are any workarounds. The lack of these functions makes dealing with nonlinear models cumbersome, even as I try to convert the problems into more tractable forms. Thanks in advance!

**Comment by odow** [comment]
You can read more about the current design of JuMP, and why nonlinear is different, in this paper: https://arxiv.org/abs/2002.03447. There are no workarounds. If you need to delete a nonlinear constraint, rebuild the model instead.

We won't be supporting `delete` before https://github.com/jump-dev/MathOptInterface.jl/issues/846 is implemented. The current timeline for this is 2-3 years, so don't wait for it.

`is_valid` is perhaps easier to implement, so I'll leave this issue open for it. Are there any others?

TODOs
- [ ] Implement `is_valid(::JuMP.Model, ::ConstraintRef{Model, NonlinearConstraintIndex, ScalarShape})`

**Comment by 1ozturkbe** [comment]
I guess this is not implemented for either ConstraintRefs or NonlinearConstraintRefs, but efficient ways to extract variables from constraints/NLconstraints would be great. Also, if you ask me, the following  MWE should not return an error...
``` julia
using JuMP, Gurobi
m = Model(Gurobi.Optimizer)
@variable(m, x)
ex = @NLexpression(m, x^2)
@constraint(m, x >=1)
@objective(m, Min, x)
optimize!(m)
```
but it does (saying ERROR: The solver does not support nonlinear problems...) although there are truly no nonlinear constraints in the model. So somehow the behavior of expressions is also different.

**Comment by odow** [comment]
> I guess the issues is that there is no equivalent of @build_constrant for nonlinear expressions, where the expressions are not attached to the model.

Correct.

**Comment by 1ozturkbe** [comment]
Well, do you think this is an issue? I see that you listed work on nonlinear functions as future work, but for folks working with JuMP the lack of these functions hamstrings attempts to interact with anything nonlinear in the meanwhile. You can't remove constraints, see their variables, or write unattached expressions. Is there at least a workaround?

**Comment by odow** [comment]
We are aware that nonlinear support is a second-class citizen in the current version of JuMP. There are no workarounds, other than structuring your work in a way that avoids these problems, e.g., by rebuilding the model instead of trying to delete a constraint.

We spent the last four years rewriting the lp/qp/conic portion of JuMP to fix issues like being unable to delete constraints. We deferred the NLP rewrite because it is a significant piece of work that we don't have developer time for at present. It's on the TODO list, but there are higher priority items like improving the documentation for a JuMP 1.0 release. Better NLP is a JuMP 2.0 feature.

**Comment by 1ozturkbe** [comment]
Alright, thanks!

**Comment by Jovansam** [comment]
> (saying ERROR: The solver does not support nonlinear problems...)

Talking of workarounds: why would you not do 
```
using JuMP, Gurobi
m = Model(Gurobi.Optimizer)
@variable(m, x)
#ex = @NLexpression(m, x^3)
@constraint(m, x >=1)
@objective(m, Min, x)
optimize!(m)
```
since you really do not use ex in the model. And that way the model returns a solution. Otherwise you would just change the @objective to @NLobjective

But of course I still look forward to future of JuMP with greater NLP support, among other features.

**Comment by 1ozturkbe** [comment]
@Jovansam I hope you realize that what I provided is MWE... I don't need a NLP to minimize x^3 😊

**Comment by Jovansam** [comment]
@1ozturkbe . Noted. But you also miss the point: you can can always find a work around until the feature is implemented. I have been in many situations like yours....

**Comment by hdavid16** [comment]
@1ozturkbe for `is_valid`, you can check, for now, if the constraint name is in the object dictionary for the model (`m.obj_dict`). This obviously requires your nonlinear constraint to be named and does not guarantee it is a valid constraint, it just checks that it is a registered name.

**Comment by odow** [comment]
> if the constraint name is in the object dictionary for the model 

Use `object_dictionary(model)` instead of `model.obj_dict`.

The `is_valid` method is something like the following, if someone wants to implement it:
```julia
function is_valid(
    model::Model, 
    c::ConstraintRef{Model,NonlinearConstraintIndex},
)
    if model !== c.model
        return false
    end
    _init_NLP(model)
    return 1 <= c.index.value <= num_nl_constraints(model)
end
```

**Comment by odow** [comment]
Closed by #2592
